File: ui/src/lib/shared/stores/metadata.ts
import { writable, get } from 'svelte/store';
import { api } from '../utils/api';
import {
	createColorHelper,
	createHomarrIconComponent,
	createIconComponent,
	createStyle,
	type ColorStyle
} from '../utils/styling';

export interface TypeMetadata<TMetadata = Record<string, unknown>> {
	id: string;
	name: string;
	description: string;
	category: string;
	icon: string;
	color: string;
	metadata: TMetadata;
}

export interface EntityMetadata {
	id: string;
	color: string;
	icon: string;
}

export interface MetadataRegistry {
	service_definitions: TypeMetadata<ServicedDefinitionMetadata>[];
	subnet_types: TypeMetadata<SubnetTypeMetadata>[];
	edge_types: TypeMetadata<EdgeTypeMetadata>[];
	group_types: TypeMetadata<GroupTypeMetadata>[];
	entities: EntityMetadata[];
	ports: TypeMetadata<PortTypeMetadata>[];
}

export interface ServicedDefinitionMetadata {
	can_be_added: boolean;
	is_dns_resolver: boolean;
	is_gateway: boolean;
	is_reverse_proxy: boolean;
	is_generic: boolean;
	manages_virtualization: 'vms' | 'containers';
	has_homarr_icon: boolean;
	layer: 'Layer3' | 'Layer4';
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface SubnetTypeMetadata {}

export interface EdgeTypeMetadata {
	is_dashed: boolean;
	has_start_marker: boolean;
	has_end_marker: boolean;
	style_label_like_nodes: boolean;
}

// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface GroupTypeMetadata {}

export interface PortTypeMetadata {
	is_management: boolean;
	is_dns: boolean;
	is_custom: boolean;
	can_be_added: boolean;
	number: number;
	protocol: 'Tcp' | 'Udp';
}

export const metadata = writable<MetadataRegistry>();

// Shared color helper functions that work for both TypeMetadata and EntityMetadata
function createSharedHelpers<T extends keyof MetadataRegistry>(category: T) {
	return {
		getColorString: (id: string | null): string => {
			const $registry = get(metadata);
			const item = $registry?.[category]?.find((item) => item.id === id);
			return item?.color || 'gray';
		},

		getColorHelper: (id: string | null): ColorStyle => {
			const $registry = get(metadata);
			const item = $registry?.[category]?.find((item) => item.id === id);
			const baseColor = item?.color || null;
			return createColorHelper(baseColor);
		},

		getIcon: (id: string | null) => {
			const $registry = get(metadata);
			return (
				($registry?.[category] as EntityMetadata[])?.find((item) => item.id === id)?.icon ||
				'HelpCircle'
			);
		},

		getIconComponent: (id: string | null) => {
			const $registry = get(metadata);
			const item = ($registry?.[category] as EntityMetadata[])?.find((item) => item.id === id);
			const iconName = item?.icon || null;
			return createIconComponent(iconName);
		},

		getStyle: (id: string | null) => {
			const $registry = get(metadata);
			const item = ($registry?.[category] as EntityMetadata[])?.find((item) => item.id === id);
			const color = item?.color || null;
			const icon = item?.icon || null;
			return createStyle(color, icon);
		}
	};
}

// Type helpers to constrain generic types
type TypeMetadataKeys = {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	[K in keyof MetadataRegistry]: MetadataRegistry[K][number] extends TypeMetadata<any> ? K : never;
}[keyof MetadataRegistry];

type EntityMetadataKeys = {
	[K in keyof MetadataRegistry]: MetadataRegistry[K][number] extends EntityMetadata ? K : never;
}[keyof MetadataRegistry];

// Full TypeMetadata helpers (includes color methods + other methods)
function createTypeMetadataHelpers<T extends TypeMetadataKeys>(category: T) {
	const sharedHelpers = createSharedHelpers(category);

	// Extract metadata type from the registry
	type MetadataType = MetadataRegistry[T][number] extends TypeMetadata<infer M> ? M : never;

	const helpers = {
		...sharedHelpers,

		getIconComponent: (id: string | null) => {
			const $registry = get(metadata);
			const item = ($registry?.[category] as TypeMetadata<MetadataType>[])?.find(
				(item) => item.id === id
			);
			const iconName = item?.icon || null;

			if (
				item?.metadata &&
				typeof item.metadata === 'object' &&
				'has_homarr_icon' in item.metadata &&
				item.metadata.has_homarr_icon
			) {
				return createHomarrIconComponent(iconName);
			} else {
				return createIconComponent(iconName);
			}
		},

		getItems: () => {
			const $registry = get(metadata);
			return $registry?.[category] as TypeMetadata<MetadataType>[];
		},

		getItem: (id: string | null) => {
			const $registry = get(metadata);
			return (
				($registry?.[category] as TypeMetadata<MetadataType>[])?.find((item) => item.id === id) ||
				null
			);
		},

		getName: (id: string | null) => {
			const $registry = get(metadata);
			return (
				($registry?.[category] as TypeMetadata<MetadataType>[])?.find((item) => item.id === id)
					?.name ||
				id ||
				''
			);
		},

		getDescription: (id: string | null) => {
			const $registry = get(metadata);
			return (
				($registry?.[category] as TypeMetadata<MetadataType>[])?.find((item) => item.id === id)
					?.description || ''
			);
		},

		getCategory: (id: string | null) => {
			const $registry = get(metadata);
			return (
				($registry?.[category] as TypeMetadata<MetadataType>[])?.find((item) => item.id === id)
					?.category || ''
			);
		},

		getMetadata: (id: string | null): MetadataType => {
			const $registry = get(metadata);
			return (
				($registry?.[category] as TypeMetadata<MetadataType>[])?.find((item) => item.id === id)
					?.metadata || ({} as MetadataType)
			);
		}
	};

	return helpers;
}

// EntityMetadata helpers (only color methods)
function createEntityMetadataHelpers<T extends EntityMetadataKeys>(category: T) {
	const sharedHelpers = createSharedHelpers(category);

	const helpers = {
		getItems: () => {
			const $registry = get(metadata);
			return $registry?.[category] as EntityMetadata[];
		},

		getItem: (id: string | null) => {
			const $registry = get(metadata);
			return ($registry?.[category] as EntityMetadata[])?.find((item) => item.id === id) || null;
		},

		// Only include the shared color methods
		...sharedHelpers
	};

	return helpers;
}

// Create all the helpers
export const serviceDefinitions = createTypeMetadataHelpers('service_definitions');
export const subnetTypes = createTypeMetadataHelpers('subnet_types');
export const edgeTypes = createTypeMetadataHelpers('edge_types');
export const groupTypes = createTypeMetadataHelpers('group_types');
export const entities = createEntityMetadataHelpers('entities');
export const ports = createTypeMetadataHelpers('ports');

export async function getMetadata() {
	await api.request<MetadataRegistry>('/metadata', metadata, (metadata) => metadata, {
		method: 'GET'
	});
}

File: server/src/server/shared/types/metadata.rs
use serde::Serialize;

#[derive(Serialize, Debug, Clone)]
pub struct MetadataRegistry {
    pub service_definitions: Vec<TypeMetadata>,
    pub subnet_types: Vec<TypeMetadata>,
    pub edge_types: Vec<TypeMetadata>,
    pub entities: Vec<EntityMetadata>,
    pub ports: Vec<TypeMetadata>
}

#[derive(Serialize, Debug, Clone)]
pub struct TypeMetadata {
    pub id: &'static str,
    pub name: Option<&'static str>,
    pub description: Option<&'static str>,
    pub category: Option<&'static str>,
    pub icon: Option<&'static str>,
    pub color: Option<&'static str>,
    pub metadata: Option<serde_json::Value>,
}

#[derive(Serialize, Debug, Clone)]
pub struct EntityMetadata {
    pub id: &'static str,
    pub color: &'static str,
    pub icon: &'static str
}

pub trait HasId {
    fn id(&self) -> &'static str;
}

pub trait MetadataProvider<T>: HasId {
    fn to_metadata(&self) -> T;
}

pub trait EntityMetadataProvider: MetadataProvider<EntityMetadata> {
    fn color(&self) -> &'static str;
    fn icon(&self) -> &'static str;
}

pub trait TypeMetadataProvider: EntityMetadataProvider + MetadataProvider<TypeMetadata> {
    fn name(&self) -> &'static str;
    fn description(&self) -> &'static str {
        ""
    }
    fn category(&self) -> &'static str {
        ""
    }
    fn metadata(&self) -> serde_json::Value {
        serde_json::json!({})
    }
}

impl<T> MetadataProvider<EntityMetadata> for T 
where 
    T: EntityMetadataProvider,
{
    fn to_metadata(&self) -> EntityMetadata {
        EntityMetadata { 
            id: self.id(), 
            color: self.color(),
            icon: self.icon()
        }
    }
}

impl<T> MetadataProvider<TypeMetadata> for T 
where 
    T: TypeMetadataProvider,
{
    fn to_metadata(&self) -> TypeMetadata {
        let id = self.id();
        let name = self.name();
        let description = self.description();
        let category = self.category();
        let icon = self.icon();
        let color = self.color();
        let metadata = self.metadata();

        TypeMetadata {
            id,
            name: (!name.is_empty()).then_some(name),
            description: (!description.is_empty()).then_some(description),
            category: (!category.is_empty()).then_some(category),
            icon: (!icon.is_empty()).then_some(icon),
            color: (!color.is_empty()).then_some(color),
            metadata: (!metadata.as_object().map_or(false, |obj| obj.is_empty())).then_some(metadata)
        }
    }
}